utils/pubsub.go

```diff
@@ -81,45 +81,39 @@ type subPublisher struct {
 	// the last value waiting to be published to the channel
 	lastValue *DataEvent
 
-	startedLock sync.Mutex
-	// flag indicating whether the publisher's internal goroutine is started or not
-	started bool
+	startedOnce sync.Once
+	ping        chan struct{}
 }
 
 // publish sets the publisher's lastValue and starts the
 // internal goroutine if it is not already started
 func (r *subPublisher) publish(data *DataEvent) {
+	r.startedOnce.Do(func() {
+		r.ping = make(chan struct{}, 1)
+		go r.startLoop()
+	})
 
 	r.lastValueLock.Lock()
+	defer r.lastValueLock.Unlock()
+
 	// update last value
 	r.lastValue = data
-	r.lastValueLock.Unlock()
 
-	r.startedLock.Lock()
-	// start publish loop if not started
-	if !r.started {
-		go r.startLoop()
-		r.started = true
+	select {
+	case r.ping <- struct{}{}: // signals the startLoop that it has to read the value
```

Comment:
please explain the behaviour here. We wouldn't want publish method to block, the consumer's channel might take a long time to process an event

Reply:
actually this is pretty smart!